src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/OrderPreferences/GeneralType.php

```diff
@@ -53,16 +57,35 @@ class GeneralType extends TranslatorAwareType
      */
     private $tosCmsChoices;
 
+    /**
+     * @var \PrestaShopBundle\Service\Routing\Router
+     */
+    private $router;
+
     public function __construct(
         TranslatorInterface $translator,
         array $locales,
         $defaultCurrencyIsoCode,
-        array $tosCmsChoices
+        array $tosCmsChoices,
+        Router $router
     ) {
         parent::__construct($translator, $locales);
 
         $this->defaultCurrencyIsoCode = $defaultCurrencyIsoCode;
         $this->tosCmsChoices = $tosCmsChoices;
+        $this->router = $router;
```

Comment:
I don't like router inside form type.
btw if Im not mistake, each form action can be modified when its being built, so it might not be necessary to set the action in form type.

Reply:
Agreed with @zuk3975 : the FormType aims to be a class that carries the definition of the form. What inputs it has, what constraints it has. It should not, in an ideal world*, carry data about "how to render" the form and the app/view layer stuff.

We might have mixed the FormType and app/view layer stuff but either it was a mistake or else there must be a good reason.

*Something that never exists 😅 

<hr>

So indeed we should "give to the Type the last few things needed to be 100% renderable by Twig Form Theme" somewhere else.

Maybe we could pass the Type to another class, in charge of this. Maybe we could decorate it ?